packages/doenetml-worker-rust/lib-doenetml-core/src/core/components/component_builder.rs

```diff
@@ -0,0 +1,361 @@
+    pub fn from_normalized_root(normalized_root: &NormalizedRoot) -> Self {
+        let mut builder = Self::from_normalized_root_without_extending(normalized_root);
+        for idx in 0..builder.components.len() {
+            if !(matches!(&normalized_root.nodes[idx], NormalizedNode::Element(_))) {
+                continue;
+            };
+            let elm = match &normalized_root.nodes[idx] {
+                NormalizedNode::Element(elm) => elm,
+                _ => unreachable!(),
+            };
+            let extending = elm.extending.clone().map(|e| e.take_resolution());
+            if extending.is_none() {
+                continue;
+            }
+            let ref_resolution = extending.unwrap();
+            let component = &builder.components[idx];
+            let referent = &builder.components[ref_resolution.node_idx];
+
+            match Self::determine_extending(&ref_resolution, component, referent) {
+                Ok(extending) => {
+                    builder.components[idx].set_extending(extending);
+                }
+                Err(err) => {
+                    builder.components[idx] = ComponentEnum::_Error(_Error {
+                        message: format!("Error while extending: {}", err),
+                        ..Default::default()
+                    });
+                }
+            }
+        }
+        builder
+    }
+
+    /// Creates all `components` but sets all their `extending` fields to `None`.
+    /// This is an intermediate step that needs to be done before resolving references in `extending`.
+    fn from_normalized_root_without_extending(normalized_root: &NormalizedRoot) -> Self {
+        // We are going to create components possibly out of order. We will track which components are created
+        // and which are in the process of being created.
+        let mut components: Vec<Option<ComponentEnum>> = iter::repeat_with(|| None)
+            .take(normalized_root.nodes.len())
+            .collect();
+        #[derive(Clone, Copy, Debug)]
+        enum Status {
+            /// No component has been created at this index.
+            Empty,
+            /// A component is in the process of being created at this index.
+            Creating,
+            /// A component has been created at this index.
+            Created,
+        }
+        let mut component_status: Vec<Status> = iter::repeat(Status::Empty)
+            .take(normalized_root.nodes.len())
+            .collect();
+
+        // Creating the nodes lowest-index first should lead to less queueing than the other way around.
+        let mut queue: Vec<usize> = Vec::from_iter((0..components.len()).rev());
+
+        while let Some(idx) = queue.pop() {
+            if components[idx].is_some() {
+                // No need to remake a component.
+                continue;
+            }
+            if matches!(component_status[idx], Status::Creating) {
```

Comment:
The component status currently isn't doing anything. The only check is here where it checks for `Creating`, but it can never happen given that the status is always changed to `Created` or `Empty` before continuing the loop. We can just remove it.

Reply:
For this reason, this check does not catch circular dependencies. We get "maximum call stack size" errors with either of these two DoenetMLs:
```
<text name="t1" extend="$t1" />
```
or
```
<text name="t1" extend="$t2" />
<text name="t2" extend="$t1" />
```

Reply:
Of course, the response to this circular dependency is arguably worse:
```
<text name="t1" extend="$t1.value" />
```
We don't run into trouble with this code, but the code for freshening dependencies gets into infinite recursion. Since we designed that code to handle a large stack, it grinds on for a long time before finally panicking.

Reply:
Hmmm...okay. I'll check another way.

Reply:
Even though I can't trigger a circular dependence in this code, the check seems reasonable for when we add features that could lead to a circular dependence here.